### DISCUSSION TITLE
Should optimisticData function treat it's parameter as immutable?

### DISCUSSION THREAD

**Discussion by ShakedHad** [vercel/swr#2730] [Q&A]
Hi
I have a question about using optimisticData function, as it looks from this [docs](https://swr.vercel.app/docs/mutation#optimistic-updates), the currentData parameter should be treated as immutable, creating new object upon return (same as Redux's reducers).
But after some tests i found that even when treating it as mutable everything still works fine, moreover, when i tried to implement object cloning (for immutability) things starts to break and sometime cause the optimisticData no to work at all.

For example, the key `/examples` saves a list of object of type `{id:1, exmple: true}`.

how should i implement an optimisticData function:
immutably: 
```
optimisticData: (data: Example[]) => {
    return [{...data[0], example: false},...data.slice(1)];
}
```
or mutably:
```
optimisticData: (data: Example[]) => {
    data[0].example = false;
    return data;
}
```

Also, if immutably is the right approach, should a shallow copy be suffice? or do i need to implement something like `_.deepClone()`?

**Comment by boris-arkenaar** [comment]
My experience is: you should not mutate it, but a shallow clone suffices. Don't remember exactly what the effect was of mutating.
